common/js/src/logging/logging.js

```diff
@@ -314,24 +317,45 @@ function setupRootLoggerLevel() {
 }
 
 function setupLogBuffer() {
-  /** @type {!GSC.LogBuffer} */
-  let logBuffer;
-  if (goog.object.containsKey(
-          window, GSC.Logging.GLOBAL_LOG_BUFFER_VARIABLE_NAME)) {
-    logBuffer = window[GSC.Logging.GLOBAL_LOG_BUFFER_VARIABLE_NAME];
-    goog.log.fine(
-        logger,
-        'Detected an existing log buffer instance, attaching it to ' +
-            'the root logger');
-  } else {
-    logBuffer = new GSC.LogBuffer(LOG_BUFFER_CAPACITY);
-    window[GSC.Logging.GLOBAL_LOG_BUFFER_VARIABLE_NAME] = logBuffer;
-    goog.log.fine(
-        logger,
-        'Created a new log buffer instance, attaching it to the root logger');
-  }
-
-  logBuffer.attachToLogger(rootLogger, document.location.href);
+  GSC.LogBuffer.attachBufferToLogger(
+      logBuffer, rootLogger, document.location.href);
+
+  // Expose our log buffer in the global window properties. Pages other than the
+  // background will use it to access the background page's log buffer - see the
+  // code directly below.
+  goog.global[GLOBAL_LOG_BUFFER_VARIABLE_NAME] = logBuffer;
+
+  chrome.runtime.getBackgroundPage(function(backgroundPage) {
+    GSC.Logging.check(backgroundPage);
+    goog.asserts.assert(backgroundPage);
+
+    if (backgroundPage === window) {
+      // We're running inside the background page - no action needed.
+      return;
+    }
+
+    // We've discovered we're running outside the background page - so need to
+    // switch to using the background page's log buffer in order to keep all
+    // logs aggregated and available in a single place.
+    // First, obtain a reference to the background page's log buffer.
+    const backgroundLogBuffer =
+        /** @type {GSC.LogBuffer} */ (
+            backgroundPage[GLOBAL_LOG_BUFFER_VARIABLE_NAME]);
+    GSC.Logging.check(backgroundLogBuffer);
+    goog.asserts.assert(backgroundLogBuffer);
+    // Copy the logs we've accumulated in the current page into the background
+    // page's log buffer.
+    logBuffer.copyToOtherBuffer(backgroundLogBuffer);
+    // From now, start using the background page's buffer for collecting data
+    // from our page's loggers. Dispose of our log buffer to avoid storing new
+    // logs twice.
+    GSC.LogBuffer.attachBufferToLogger(
+        backgroundLogBuffer, rootLogger, document.location.href);
+    logBuffer.dispose();
+    // Switch all references to the background page's log buffer.
+    logBuffer = backgroundLogBuffer;
+    goog.global[GLOBAL_LOG_BUFFER_VARIABLE_NAME] = logBuffer;
```

Comment:
Is this ever accessed if we are not the background page? If yes, why not use getLogBuffer() instead? If not, can we just set this to null?

Reply:
Done - removed this property, as it's not needed indeed.